File: src/mcp_acp/api/schemas/incidents.py
"""Incidents API schemas."""

from __future__ import annotations

__all__ = [
    "IncidentsSummary",
]

from pydantic import BaseModel, Field


class IncidentsSummary(BaseModel):
    """Summary of incidents for badge state and overview.

    Provides counts and latest timestamps for determining unread state.

    Critical incidents (contribute to badge):
    - Shutdowns: Intentional security shutdowns (audit failure, session hijacking)
    - Emergency: Audit fallback entries when normal audit fails

    Informational (don't contribute to badge):
    - Bootstrap: Startup validation errors (config/policy issues)
    """

    shutdowns_count: int = Field(description="Number of security shutdown entries")
    emergency_count: int = Field(description="Number of emergency audit entries")
    bootstrap_count: int = Field(description="Number of bootstrap error entries")

    # Latest timestamp for critical incidents only (for badge calculation)
    # Bootstrap is informational and doesn't trigger badge
    latest_critical_timestamp: str | None = Field(
        default=None,
        description="ISO timestamp of most recent shutdown or emergency entry",
    )

    # File paths for debugging
    shutdowns_path: str | None = Field(default=None, description="Path to shutdowns.jsonl")
    emergency_path: str | None = Field(default=None, description="Path to emergency_audit.jsonl")
    bootstrap_path: str | None = Field(default=None, description="Path to bootstrap.jsonl")

File: web/src/api/incidents.ts
/**
 * Incidents API client.
 *
 * Provides access to incident history:
 * - Shutdowns: Intentional security shutdowns (audit failure, session hijacking)
 * - Bootstrap: Startup validation errors (config/policy issues)
 * - Emergency: Audit fallback entries when normal audit fails
 *
 * Uses manager-level API: /api/manager/incidents - aggregated from all proxies
 */

import { apiGet, type RequestOptions } from './client'
import type { AggregatedIncidentsResponse } from '@/types/api'

// =============================================================================
// Types
// =============================================================================

export interface IncidentsSummary {
  shutdowns_count: number
  emergency_count: number
  bootstrap_count: number
  latest_critical_timestamp: string | null
  shutdowns_path: string | null
  emergency_path: string | null
  bootstrap_path: string | null
}

export type IncidentType = 'shutdown' | 'bootstrap' | 'emergency'

// =============================================================================
// Manager-level API (aggregated from all proxies)
// =============================================================================

export interface AggregatedIncidentsParams {
  proxy?: string
  incident_type?: IncidentType
  time_range?: string
  limit?: number
  before?: string
}

/**
 * Get aggregated incidents from all proxies.
 * Combines shutdowns (per-proxy) with bootstrap and emergency (global).
 * Each entry includes 'incident_type' and 'proxy_name' (for shutdowns).
 */
export async function getAggregatedIncidents(
  params: AggregatedIncidentsParams = {},
  options?: RequestOptions
): Promise<AggregatedIncidentsResponse> {
  const searchParams = new URLSearchParams()
  if (params.proxy) searchParams.set('proxy', params.proxy)
  if (params.incident_type) searchParams.set('incident_type', params.incident_type)
  if (params.time_range) searchParams.set('time_range', params.time_range)
  if (params.limit) searchParams.set('limit', String(params.limit))
  if (params.before) searchParams.set('before', params.before)

  const query = searchParams.toString()
  const url = query ? `/manager/incidents?${query}` : '/manager/incidents'
  return apiGet<AggregatedIncidentsResponse>(url, options)
}

/**
 * Get incidents summary with counts and latest timestamp.
 * Used for badge state calculation.
 *
 * Uses manager-level endpoint to aggregate from all proxies.
 *
 * @param params.since - Only count entries after this ISO timestamp (for unread badge).
 */
export async function getIncidentsSummary(
  params?: { since?: string },
  options?: RequestOptions,
): Promise<IncidentsSummary> {
  const searchParams = new URLSearchParams()
  if (params?.since) searchParams.set('since', params.since)
  const query = searchParams.toString()
  const url = query ? `/manager/incidents/summary?${query}` : '/manager/incidents/summary'
  return apiGet<IncidentsSummary>(url, options)
}
